dm/master/shardddl/optimist.go

```diff
@@ -0,0 +1,469 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shardddl
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"go.etcd.io/etcd/clientv3"
+	"go.uber.org/zap"
+
+	"github.com/pingcap/dm/pkg/etcdutil"
+	"github.com/pingcap/dm/pkg/log"
+	"github.com/pingcap/dm/pkg/shardddl/optimism"
+)
+
+// Optimist is used to coordinate the shard DDL migration in optimism mode.
+type Optimist struct {
+	mu sync.Mutex
+
+	logger log.Logger
+
+	closed bool
+	cancel context.CancelFunc
+	wg     sync.WaitGroup
+
+	cli *clientv3.Client
+	lk  *optimism.LockKeeper
+	tk  *optimism.TableKeeper
+}
+
+// NewOptimist creates a new Optimist instance.
+func NewOptimist(pLogger *log.Logger) *Optimist {
+	return &Optimist{
+		logger: pLogger.WithFields(zap.String("component", "shard DDL optimist")),
+		closed: true,
+		lk:     optimism.NewLockKeeper(),
+		tk:     optimism.NewTableKeeper(),
+	}
+}
+
+// Start starts the shard DDL coordination in optimism mode.
+func (o *Optimist) Start(pCtx context.Context, etcdCli *clientv3.Client) error {
+	o.logger.Info("the shard DDL optimist is starting")
+
+	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	o.cli = etcdCli // o.cli should be set before watching and recover locks because these operations need o.cli
+
+	revSource, revInfo, revOperation, err := o.rebuildLocks(etcdCli)
+	if err != nil {
+		return err
+	}
+
+	ctx, cancel := context.WithCancel(pCtx)
+
+	o.wg.Add(1)
+	go func() {
+		defer o.wg.Done()
+		// TODO: handle fatal error from run
+		o.run(ctx, etcdCli, revSource, revInfo, revOperation)
```

Comment:
ditto

Reply:
addressed in https://github.com/pingcap/dm/pull/568/commits/9e18ddb87a498f521d446eb1937a70528b4a6856.